File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/common/LiveDataImpl.kt
package dev.zeekr.camerax_android.common

import androidx.camera.core.CameraState
import androidx.camera.core.ZoomState
import androidx.camera.view.PreviewView
import androidx.camera.view.TapToFocusInfo
import androidx.lifecycle.LiveData
import dev.zeekr.camerax_android.CameraStateApi
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.LowLightBoostStateApi
import dev.zeekr.camerax_android.PigeonApiCameraStateLiveDataProxyApi
import dev.zeekr.camerax_android.PigeonApiIntLiveDataProxyApi
import dev.zeekr.camerax_android.PigeonApiLowLightBoostStateLiveDataProxyApi
import dev.zeekr.camerax_android.PigeonApiPreviewViewStreamStateLiveDataProxyApi
import dev.zeekr.camerax_android.PigeonApiTapToFocusInfoLiveDataProxyApi
import dev.zeekr.camerax_android.PigeonApiTorchStateLiveDataProxyApi
import dev.zeekr.camerax_android.PigeonApiZoomStateLiveDataProxyApi
import dev.zeekr.camerax_android.PreviewViewStreamStateApi
import dev.zeekr.camerax_android.TorchStateApi
import dev.zeekr.camerax_android.core.api
import dev.zeekr.camerax_android.core.lowLightBoostStateApi
import dev.zeekr.camerax_android.core.torchStateApi
import dev.zeekr.camerax_android.view.api

class IntLiveDataImpl(registrar: CameraXApiPigeonProxyApiRegistrar) : PigeonApiIntLiveDataProxyApi(registrar) {
    override fun getValue(pigeon_instance: IntLiveData): Long? {
        return pigeon_instance.instance.value?.toLong()
    }

    override fun observeForever(pigeon_instance: IntLiveData, observer: IntObserver) {
        return pigeon_instance.instance.observeForever(observer)
    }

    override fun removeObserver(pigeon_instance: IntLiveData, observer: IntObserver) {
        return pigeon_instance.instance.removeObserver(observer)
    }
}

class CameraStateLiveDataImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiCameraStateLiveDataProxyApi(registrar) {
    override fun getValue(pigeon_instance: CameraStateLiveData): CameraStateApi? {
        return pigeon_instance.instance.value?.api
    }

    override fun observeForever(pigeon_instance: CameraStateLiveData, observer: CameraStateObserver) {
        pigeon_instance.instance.observeForever(observer)
    }

    override fun removeObserver(pigeon_instance: CameraStateLiveData, observer: CameraStateObserver) {
        return pigeon_instance.instance.removeObserver(observer)
    }
}

class TorchStateLiveDataImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiTorchStateLiveDataProxyApi(registrar) {
    override fun getValue(pigeon_instance: TorchStateLiveData): TorchStateApi? {
        return pigeon_instance.instance.value?.torchStateApi
    }

    override fun observeForever(pigeon_instance: TorchStateLiveData, observer: TorchStateObserver) {
        pigeon_instance.instance.observeForever(observer)
    }

    override fun removeObserver(pigeon_instance: TorchStateLiveData, observer: TorchStateObserver) {
        pigeon_instance.instance.removeObserver(observer)
    }
}

class ZoomStateLiveDataImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiZoomStateLiveDataProxyApi(registrar) {
    override fun getValue(pigeon_instance: ZoomStateLiveData): ZoomState? {
        return pigeon_instance.instance.value
    }

    override fun observeForever(pigeon_instance: ZoomStateLiveData, observer: ZoomStateObserver) {
        return pigeon_instance.instance.observeForever(observer)
    }

    override fun removeObserver(pigeon_instance: ZoomStateLiveData, observer: ZoomStateObserver) {
        return pigeon_instance.instance.removeObserver(observer)
    }
}

class TapToFocusInfoLiveDataImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiTapToFocusInfoLiveDataProxyApi(registrar) {
    override fun getValue(pigeon_instance: TapToFocusInfoLiveData): TapToFocusInfo? {
        return pigeon_instance.instance.value
    }

    override fun observeForever(pigeon_instance: TapToFocusInfoLiveData, observer: TapToFocusInfoObserver) {
        return pigeon_instance.instance.observeForever(observer)
    }

    override fun removeObserver(pigeon_instance: TapToFocusInfoLiveData, observer: TapToFocusInfoObserver) {
        return pigeon_instance.instance.removeObserver(observer)
    }
}

class LowLightBoostStateLiveDataImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiLowLightBoostStateLiveDataProxyApi(registrar) {
    override fun getValue(pigeon_instance: LowLightBoostStateLiveData): LowLightBoostStateApi? {
        return pigeon_instance.instance.value?.lowLightBoostStateApi
    }

    override fun observeForever(pigeon_instance: LowLightBoostStateLiveData, observer: LowLightBoostStateObserver) {
        pigeon_instance.instance.observeForever(observer)
    }

    override fun removeObserver(pigeon_instance: LowLightBoostStateLiveData, observer: LowLightBoostStateObserver) {
        pigeon_instance.instance.removeObserver(observer)
    }
}

class PreviewViewStreamStateLiveDataImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiPreviewViewStreamStateLiveDataProxyApi(registrar) {
    override fun getValue(pigeon_instance: PreviewViewStreamStateLiveData): PreviewViewStreamStateApi? {
        return pigeon_instance.instance.value?.api
    }

    override fun observeForever(
        pigeon_instance: PreviewViewStreamStateLiveData, observer: PreviewViewStreamStateObserver
    ) {
        return pigeon_instance.instance.observeForever(observer)
    }

    override fun removeObserver(
        pigeon_instance: PreviewViewStreamStateLiveData, observer: PreviewViewStreamStateObserver
    ) {
        return pigeon_instance.instance.removeObserver(observer)
    }
}

class IntLiveData(val instance: LiveData<Int>)
class CameraStateLiveData(val instance: LiveData<CameraState>)
class TorchStateLiveData(val instance: LiveData<Int>)
class ZoomStateLiveData(val instance: LiveData<ZoomState>)
class TapToFocusInfoLiveData(val instance: LiveData<TapToFocusInfo>)
class LowLightBoostStateLiveData(val instance: LiveData<Int>)
class PreviewViewStreamStateLiveData(val instance: LiveData<PreviewView.StreamState>)

File: camerax_android/lib/src/common/live_data_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/common.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_android/src/view.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

abstract base class LiveDataImpl<T> implements LiveData<T> {}

final class IntLiveDataImpl extends LiveDataImpl<int> {
  final IntLiveDataProxyApi api;

  IntLiveDataImpl.internal(this.api);

  @override
  Future<int?> getValue() => api.getValue();

  @override
  Future<void> observeForever(Observer<int> observer) =>
      api.observeForever(observer.api);

  @override
  Future<void> removeObserver(Observer<int> observer) =>
      api.removeObserver(observer.api);
}

final class CameraStateLiveDataImpl extends LiveDataImpl<CameraState> {
  final CameraStateLiveDataProxyApi api;

  CameraStateLiveDataImpl.internal(this.api);

  @override
  Future<CameraState?> getValue() => api.getValue().then((e) => e?.impl);

  @override
  Future<void> observeForever(Observer<CameraState> observer) =>
      api.observeForever(observer.api);

  @override
  Future<void> removeObserver(Observer<CameraState> observer) =>
      api.removeObserver(observer.api);
}

final class TorchStateLiveDataImpl extends LiveDataImpl<TorchState> {
  final TorchStateLiveDataProxyApi api;

  TorchStateLiveDataImpl.internal(this.api);

  @override
  Future<TorchState?> getValue() => api.getValue().then((e) => e?.impl);

  @override
  Future<void> observeForever(Observer<TorchState> observer) =>
      api.observeForever(observer.api);

  @override
  Future<void> removeObserver(Observer<TorchState> observer) =>
      api.removeObserver(observer.api);
}

final class ZoomStateLiveDataImpl extends LiveDataImpl<ZoomState> {
  final ZoomStateLiveDataProxyApi api;

  ZoomStateLiveDataImpl.internal(this.api);

  @override
  Future<ZoomState?> getValue() => api.getValue().then((e) => e?.impl);

  @override
  Future<void> observeForever(Observer<ZoomState> observer) =>
      api.observeForever(observer.api);

  @override
  Future<void> removeObserver(Observer<ZoomState> observer) =>
      api.removeObserver(observer.api);
}

final class TapToFocusInfoLiveDataImpl extends LiveDataImpl<TapToFocusInfo> {
  final TapToFocusInfoLiveDataProxyApi api;

  TapToFocusInfoLiveDataImpl.internal(this.api);

  @override
  Future<TapToFocusInfo?> getValue() => api.getValue().then((e) => e?.impl);

  @override
  Future<void> observeForever(Observer<TapToFocusInfo> observer) =>
      api.observeForever(observer.api);

  @override
  Future<void> removeObserver(Observer<TapToFocusInfo> observer) =>
      api.removeObserver(observer.api);
}

final class LowLightBoostStateLiveDataImpl
    extends LiveDataImpl<LowLightBoostState> {
  final LowLightBoostStateLiveDataProxyApi api;

  LowLightBoostStateLiveDataImpl.internal(this.api);

  @override
  Future<LowLightBoostState?> getValue() => api.getValue().then((e) => e?.impl);

  @override
  Future<void> observeForever(Observer<LowLightBoostState> observer) =>
      api.observeForever(observer.api);

  @override
  Future<void> removeObserver(Observer<LowLightBoostState> observer) =>
      api.removeObserver(observer.api);
}

extension IntLiveDataProxyApiX on IntLiveDataProxyApi {
  LiveData<int> get impl => IntLiveDataImpl.internal(this);
}

extension CameraStateLiveDataProxyApiX on CameraStateLiveDataProxyApi {
  LiveData<CameraState> get impl => CameraStateLiveDataImpl.internal(this);
}

extension TorchStateLiveDataProxyApiX on TorchStateLiveDataProxyApi {
  LiveData<TorchState> get impl => TorchStateLiveDataImpl.internal(this);
}

extension ZoomStateLiveDataProxyApiX on ZoomStateLiveDataProxyApi {
  LiveData<ZoomState> get impl => ZoomStateLiveDataImpl.internal(this);
}

extension TapToFocusInfoLiveDataProxyApiX on TapToFocusInfoLiveDataProxyApi {
  LiveData<TapToFocusInfo> get impl =>
      TapToFocusInfoLiveDataImpl.internal(this);
}

extension LowLightBoostStateLiveDataProxyApiX
    on LowLightBoostStateLiveDataProxyApi {
  LiveData<LowLightBoostState> get impl =>
      LowLightBoostStateLiveDataImpl.internal(this);
}
